### PR TITLE
Build refactoring - cbindgen

### PR DESCRIPTION
**Pull request by r1viollet**
# What does this PR do?

Ensure cbindgen is included in the package's build steps. By removing logics from the shell scripts, we can make build steps more modular.
This is step 1 of moving towards a common "feature" driven build system.

# Motivation

You can refer to the [RFC](https://github.com/DataDog/libdatadog/blob/main/docs/RFCs/delivering-different-features.md) around features

# Additional Notes

Anything else we should know when reviewing?

# How to test the change?

Describe here in detail how the change can be validated.

## For Reviewers
- [ ] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

File: build-telemetry-ffi.sh [hunk 1]
```diff
@@ -68,7 +68,7 @@ export RUSTFLAGS="${RUSTFLAGS:- -C relocation-model=pic}"
 
 datadog_telemetry_ffi="ddtelemetry-ffi"
 echo "Building the ${datadog_telemetry_ffi} crate (may take some time)..."
-cargo build --package="${datadog_telemetry_ffi}" --release --target "${target}"
+DESTDIR="$destdir" cargo build --package="${datadog_telemetry_ffi}" --release --target "${target}"
```
Comment (r1viollet) [inline build-telemetry-ffi.sh:71]: I am trying to debate if this is the right way of doing things.
For now we some `cp` underneath and this different mechanism.

I don't really see the benefit from the `cp` operations bellow.

Comment (r1viollet) [reply build-telemetry-ffi.sh:71]: The `OUT_DIR` set by cargo is sadly not the right place.
I think this is still slightly better, but not perfect.

File: ddcommon-ffi/build.rs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+use std::env;
+use std::path::PathBuf;
+
+fn main() {
+    let crate_dir = PathBuf::from(env::var_os("CARGO_MANIFEST_DIR").unwrap());
+    let header_name = "common.h";
+    let output_base_dir = env::var("DESTDIR").ok(); // Use `ok()` to convert Result to Option
+
+    generate_header(crate_dir, header_name, output_base_dir.as_deref());
+    println!("cargo:rerun-if-env-changed=DESTDIR");
```
Comment (danielsn) [inline ddcommon-ffi/build.rs:15]: We have the exact same code for all three crates, with the only difference being the header name.  Maybe worth factoring into a function?

File: build-common/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    let cargo_target_dir = output_base_dir.unwrap_or("target");
+
+    // Determine if `cargo_target_dir` is absolute or relative
+    let cargo_target_path = Path::new(cargo_target_dir);
+    let output_path = if cargo_target_path.is_absolute() {
+        // If absolute, use it directly
+        cargo_target_path.join("include/datadog/").join(header_name)
+    } else {
+        // If relative, adjust the path accordingly. we are in a crate, so get back to top level
+        let adjusted_path = Path::new("..").join(cargo_target_path);
```
Comment (danielsn) [inline build-common/src/lib.rs:25]: is this guaranteed to work? What if the command is run deeper in the crate?

File: ddcommon-ffi/build.rs (hunk 2 above)
Comment (r1viollet) [reply ddcommon-ffi/build.rs]: I have mixed feelings about hiding some of this behaviour :-)
I added the magical function all the same.

File: build-common/src/lib.rs (hunk 3 above)
Comment (r1viollet) [reply build-common/src/lib.rs:39]: No.
However I can not think of a way to magically infer where the headers should be placed.
I looked into using the OUT_DIR variable, but that also has different depths depending on build parameters.
So I think this being not perfect still guarantees that it works in the use cases we have today.
Happy to reconsider if you find a different logic that works well.

Comment (danielsn) [reply build-common/src/lib.rs:39]: If we want the root of the workspace, can we try https://doc.rust-lang.org/cargo/commands/cargo-locate-project.html#options

**Review by morrisonlevi — APPROVED** [review]

File: build-common/src/lib.rs (hunk 3 above)
Comment (r1viollet) [reply build-common/src/lib.rs]: Here is a new iteration using your proposal.

File: build-common/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present Datadog, Inc.
+use cbindgen::Config;
+use std::path::PathBuf;
+use std::process::Command;
+use std::str;
+use std::{env, fs};
+
+/// Configure the header generation using environment variables.
+/// call into `generate_header` with the appropriate arguments.
+///
```
Comment (r1viollet) [inline build-common/src/lib.rs:11]: I can still find ways to break this, but I feel we already improve the state of the build.

**Review by bantonsson — APPROVED** [review]

**Review by danielsn — APPROVED** [review]

**Pull request merged**